typescript/src/utils.ts

```diff
@@ -312,3 +312,55 @@ export const matchParents: MatchParentsType = (node, treeToCompare) => {
     }
     return node as any
 }
+
+export const getPositionHighlights = (position: number, sourceFile: ts.SourceFile, languageService: ts.LanguageService) => {
+    const highlights = languageService.getDocumentHighlights(sourceFile.fileName, position, [sourceFile.fileName])
+
+    if (!highlights) return
+
+    return highlights.flatMap(({ highlightSpans }) => highlightSpans.map(({ textSpan }) => textSpan.start))
+}
+
+export const isValidInitializerForDestructure = (match: ts.Expression) => {
+    const isFinalChainElement = (node: ts.Node) =>
+        ts.isThisTypeNode(node) || ts.isIdentifier(node) || ts.isParenthesizedExpression(node) || ts.isObjectLiteralExpression(node) || ts.isNewExpression(node)
+
+    const isValidChainElement = (node: ts.Node) =>
+        (ts.isPropertyAccessExpression(node) || ts.isElementAccessExpression(node) || ts.isCallExpression(node) || ts.isNonNullExpression(node)) &&
+        !ts.isOptionalChain(node)
+
+    let currentChainElement = match
+
+    while (!isFinalChainElement(currentChainElement)) {
+        if (!isValidChainElement(currentChainElement)) {
+            return false
+        }
+        type PossibleChainElement =
+            | ts.PropertyAccessExpression
+            | ts.CallExpression
+            | ts.ElementAccessExpression
+            | ts.NonNullExpression
+            | ts.ParenthesizedExpression
+            | ts.AwaitExpression
+
+        const chainElement = currentChainElement as PossibleChainElement
+
+        currentChainElement = chainElement.expression
+    }
+
+    return true
+}
+export const isNameUniqueAtLocation = (name: string, location: ts.Node | undefined, typeChecker: ts.TypeChecker) => {
+    const checker = getFullTypeChecker(typeChecker)
+
+    const newVariable = checker.resolveName(name, location as unknown as import('typescript-full').Node, ts.SymbolFlags.Value, true)
+    return !newVariable
+}
+export const isNameUniqueAtNodeClosestScope = (name: string, node: ts.Node, typeChecker: ts.TypeChecker) => {
+    const closestScope = findClosestParent(
```

Comment:
Not sure if this is the safest way to get the closest scope where variables should be unique but it works in my cases.

Reply:
This is fine I guess, but remember that in es6+ blocks (`{}`) create new scope

Reply:
fixed + added test